### PR TITLE
Added detection of custom FETCH and UID FETCH commands for IMAP

### PR DESCRIPTION
**Pull request by MonkeybreadSoftware**
Added detection of custom FETCH and UID FETCH commands for IMAP

### CODE REVIEW

**Comment by MonkeybreadSoftware** [comment]
See https://github.com/curl/curl/pull/8659 for more details.

**Comment by MonkeybreadSoftware** [comment]
To explain this. With current curl 7.82.0, we get this:

```
cs@MyMac curl-7.82.0 % /Users/cs/Temp/curl-7.82.0/src/.libs/curl -u "test@domain.com:password" -X "UID FETCH 107 BODY.PEEK[]" "imap://sslin.df.eu/INBOX" 
* 2 FETCH (UID 107 BODY[] {2211}
```

but we like to have this:

```
cs@MyMac curl-7.82.0 % curl -u "test@domain.com:password" "imap://sslin.df.eu/INBOX;UID=107"                 
Delivery-date: Wed, 30 Jan 2019 16:36:03 +0100
Received: from [80.67.18.7] (helo=mx07.ispgateway.de)
...
```

so the output of the transaction is the source code of the email and not the list line. A small change, but if you write an IMAP email application, it makes a huge difference.

File: lib/imap.c [hunk 1]
```diff
@@ -663,8 +663,34 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
-  if(!result)
-    imap_state(data, IMAP_LIST);
+  if(!result) {
+    bool isFetch = false;
+
+    /* Switch state to be FETCH if custom command is a fetch command */
+    /* so output is handled correctly. */
+
+    if(imap->custom) {
+      if(strncasecompare(imap->custom, "FETCH", 5) == 1) {
```
Comment (bagder) [inline lib/imap.c:673]:
```suggestion
      if(strncasecompare(imap->custom, "FETCH", 5)) {
```

File: lib/imap.c [hunk 2]
```diff
@@ -663,8 +663,34 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+        imap_state(data, IMAP_FETCH);
+        isFetch = true;
+      }
+
+      /* e.g. "UID FETCH 107 BODY.PEEK[]" */
+
+      if(!isFetch && imap->custom_params) {
+        if(strncasecompare(imap->custom, "UID", 3) == 1) {
```
Comment (bagder) [inline lib/imap.c:681]:
```suggestion
        if(strncasecompare(imap->custom, "UID", 3)) {
```

File: lib/imap.c [hunk 3]
```diff
@@ -663,8 +663,34 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          if(strncasecompare(imap->custom_params, " FETCH", 6) == 1) {
```
Comment (bagder) [inline lib/imap.c:682]:
```suggestion
          if(strncasecompare(imap->custom_params, " FETCH", 6)) {
```

File: lib/imap.c (hunk 1 above)
Comment (bagder) [reply lib/imap.c:673]: ... as we don't compare with 1 for this kind of boolean functions anywhere.

**Comment by bagder** [comment]
> Basically whether CURL makes a "FETCH" and you make a "FETCH" (or UID FETCH) via custom request, both are handled the same.

Why? If it already can do the transfer correctly without a custom request, why do you need to add support for doing the same transfer in a different way?

File: lib/imap.c [hunk 4]
```diff
@@ -663,8 +663,34 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
     free(mailbox);
   }
 
-  if(!result)
-    imap_state(data, IMAP_LIST);
+  if(!result) {
+    bool isFetch = false;
```
Comment (bagder) [inline lib/imap.c:667]:
```suggestion
    bool isFetch = FALSE;
```

Comment (bagder) [reply lib/imap.c:667]: We use `TRUE` and `FALSE` everywhere.

**Comment by MonkeybreadSoftware** [comment]
We need a little adjustment.

Such a request to list emails need to get LIST state, while fetch of email then needs FETCH state:

"UID FETCH 1:* (FLAGS INTERNALDATE RFC822.SIZE BODY.PEEK[HEADER.FIELDS (Message-Id DATE FROM SUBJECT TO SENDER REPLY-TO CC BCC)])"

**Comment by MonkeybreadSoftware** [comment]
I added check to distinguish between a fetch of one email vs. a listing of a range of emails.

File: tests/test1665 [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+<verify>
+<protocol>
+A001 CAPABILITY
+A002 LOGIN user secret
+A003 SELECT %TESTNUMBER
+A004 FETCH 123 BODY[1]
+A005 LOGOUT
+</protocol>
+</verify>
+</testcase>
```
Comment (bagder) [inline tests/test1665:46]: This is in the wrong directory. Move it into `data/`.

Comment (MonkeybreadSoftware) [reply tests/test1665]: Done. Sorry for all my mistakes here 😳

File: lib/imap.c [hunk 6]
```diff
@@ -663,8 +663,57 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          SUBJECT TO SENDER REPLY-TO CC BCC)])
+          to get the list of emails
+        */
+        const char *positionColon   = strchr(imap->custom_params, ':');
+        const char *positionBracket = strchr(imap->custom_params, '(');
+
+        if(positionColon) {
+          if(!positionBracket || positionBracket > positionColon) {
+            isFetch = FALSE;
+          }
```
Comment (bagder) [inline lib/imap.c:705]:
```suggestion
        if(positionColon) {
          const char *paren = strchr(imap->custom_params, '(');
          if(!paren || paren > positionColon)
            isFetch = FALSE;
```

File: lib/imap.c [hunk 7]
```diff
@@ -663,8 +663,52 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          otherwise FETCH state, e.g.
+          FETCH 175 BODY.PEEK[]
+          to get one email
+          vs.
+          UID FETCH 1:* (FLAGS INTERNALDATE RFC822.SIZE
+          BODY.PEEK[HEADER.FIELDS (Message-Id DATE FROM
+          SUBJECT TO SENDER REPLY-TO CC BCC)])
+          to get the list of emails
+        */
+        const char *positionColon   = strchr(imap->custom_params, ':');
```
Comment (bagder) [inline lib/imap.c:696]: The scan-build warns for a (reasonable) possible NULL dereference here.

[the remark this answers is not in the document]
Comment (MonkeybreadSoftware) [reply lib/imap.c]: I've removed it. This was a left over.

Comment (MonkeybreadSoftware) [reply lib/imap.c]: Let me put custom_params in a local variable and put in "" if it s NULL.

File: lib/imap.c [hunk 8]
```diff
@@ -663,8 +663,54 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+      }
+
+      /* e.g. "UID FETCH 107 BODY.PEEK[]" */
+
+      else if(strncasecompare(imap->custom, "UID", 3) &&
+              strncasecompare(imap->custom_params, " FETCH", 6)) {
+        isFetch = true;
+      }
+
+      if(isFetch) {
```
Comment (bagder) [inline lib/imap.c:684]:
```suggestion
      if(isFetch && imap->custom_params) {
```

File: lib/imap.c [hunk 9]
```diff
@@ -663,8 +663,54 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          to get one email
+          vs.
+          UID FETCH 1:* (FLAGS INTERNALDATE RFC822.SIZE
+          BODY.PEEK[HEADER.FIELDS (Message-Id DATE FROM
+          SUBJECT TO SENDER REPLY-TO CC BCC)])
+          to get the list of emails
+        */
+        const char *custom_params =
+          imap->custom_params ? imap->custom_params : "";
+        const char *positionColon = strchr(custom_params, ':');
```
Comment (bagder) [inline lib/imap.c:698]:
```suggestion
        const char *positionColon = strchr(imap->custom_params, ':');
```

... when combined with the change above.

File: lib/imap.c [hunk 10]
```diff
@@ -663,8 +663,52 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          SUBJECT TO SENDER REPLY-TO CC BCC)])
+          to get the list of emails
+        */
+        const char *positionColon   = strchr(imap->custom_params, ':');
+        if(positionColon) {
+          const char *paren = strchr(imap->custom_params, '(');
+          if(!paren || paren > positionColon)
+            isFetch = FALSE;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (bagder) [reply lib/imap.c:704]: I'm sure there are several other IMAP users out there, but we rarely see anything from them apart from the infrequent issues filed...

**Comment by MonkeybreadSoftware** [comment]
Thanks for little changes. Feels like we do pair programming on this.

File: lib/imap.c [hunk 11]
```diff
@@ -663,8 +663,52 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+          to get one email
+          vs.
+          UID FETCH 1:* (FLAGS INTERNALDATE RFC822.SIZE
+          BODY.PEEK[HEADER.FIELDS (Message-Id DATE FROM
+          SUBJECT TO SENDER REPLY-TO CC BCC)])
+          to get the list of emails
+        */
+        const char *positionColon = strchr(imap->custom_params, ':');
+        if(positionColon) {
+          const char *paren = strchr(custom_params, '(');
```
Comment (bagder) [inline lib/imap.c:698]:
```suggestion
          const char *paren = strchr(imap->custom_params, '(');
```

**Comment by bagder** [comment]
It seems this how makes a lot of CI jobs get stuck on test 647, which I cannot understand...

**Comment by MonkeybreadSoftware** [comment]
I've added another NULL check to prevent crash.

**Comment by bagder** [comment]
This still fails too many CI jobs.

**Comment by MonkeybreadSoftware** [comment]
I see, but I can't find a reason why test 647 fails.
Can you run this test and see if you get an error or crash report?

**Comment by bagder** [comment]
Failing CI jobs not updated since September, closing.

**Pull request closed without merging**